Support FastSpeech2 (+FastPitch)

kan-bayashi opened the pull request:
This PR adds FastSpeech2 module (#2019).

Changes:
- Add FastSpeech2 model.
- Add new arguments for TTS task.
	- `pitch_extractor`
	- `energy_extractor`
	- `pitch_normalize`
	- `energy_normalize`
- Add `Dio` for pitch extraction.
- Add `Energy` for energy exraction.
- Add `--write_collected_feats` option in `tts.sh` to speed up the training.
- Add `--tts_stats_dir` option in `tts.sh` to specify the stats dir from command line.

Note:
- Instead of quantized embedding of f0/energy, use continuous value.
- Use FastPitch-style token-averaged f0 / energy as discussed in #2019 

Todo:
- [x] Implement FastSpeech2
- [x] Add test for FastSpeech2
- [x] Implement F0 / energy extraction
- [x] Add test for f0 / energy extraction
- [x] Support in recipe level
- [x] check performance

Procedure to train fastspeech2
```bash
# 1. Train teacher model
./run.sh --train_config conf/tuning/train_tacotron2.yaml

# 2. Dump durations using teacher model with teacher forcing
./run.sh --stage 7 --train_config conf/tuning/train_tacotron2.yaml --test_sets "tr_no_dev dev eval1" --inference_args "--use_teacher_forcing true"

# 3. Calculate statistics pitch and energy features
./run.sh --stage 5 --stop-stage 5 --conf conf/tuning/train_fastspeech2.yaml --teacher_dumpdir /path/to/teacher_dumpdir --tts_stats_dir exp/tts_stats_for_fastspeech2
# To speed up the training, it is better to dump the features
./run.sh --stage 5 --stop-stage 5 --conf conf/tuning/train_fastspeech2.yaml --teacher_dumpdir /path/to/teacher_dumpdir --tts_stats_dir exp/tts_stats_for_fastspeech2 --write_collected_feats true

# 4. Train fastspeech2
./run.sh --stage 6 --conf conf/tuning/train_fastspeech2.yaml --teacher_dumpdir /path/to/teacher_dumpdir --tts_stats_dir exp/tts_stats_for_fastspeech2
```

kan-bayashi commented:
Almost done.

kan-bayashi commented:
@kamo-naoyuki Could you check, especially the following points?
- New options in `tts.sh`
- New feature extractor and normalize in `espnet_model.py` and `tts.py`

@r9y9 Could you check `fastspeech2.py` and `dio.py`?

r9y9 commented:
Sure, I will check them tomorrow.

kan-bayashi commented:
I uploaded some samples.
https://drive.google.com/drive/folders/18dZz5jP05Vh5bRRK8nT7HnmLXSeLsitm?usp=sharing

v1: baseline
v2: deep pitch predictor
v4: deep pitch predictor + stop gradient
gst: GST + v4

Compared to the FastSpeech samples, the quality is clearly improved.

sw005320 commented on espnet2/tts/feats_extract/dio.py:
```diff
@@ -0,0 +1,155 @@
+
+
+class Dio(AbsFeatsExtract):
+    """F0 estimation with dio + stonemask algortihm."""
```
- Maybe you can add more reference information about this (sorry if you put it somewhere)?
- It's better to state that this part cannot be a part of the computational graph and add TODO somewhere about the possibility of rewriting the f0 estimation like that.
- Do you think we can use it for ASR? If so, we'll test it.

kan-bayashi replied:
> Maybe you can add more reference information about this (sorry if you put it somewhere)?

Sure!

> It's better to state that this part cannot be a part of the computational graph and add TODO somewhere about the possibility of rewriting the f0 estimation like that.

Agree. I will add a description about it.

> Do you think we can use it for ASR? If so, we'll test it.

It can be if the task is relatively clean.
But in speech synthesis, we usually adjust the range of f0 for each speaker.
In the case of ASR, it is difficult. So I'm not sure what is happened.

sw005320 replied:
> > Do you think we can use it for ASR? If so, we'll test it.
> 
> It can be if the task is relatively clean.
> But in speech synthesis, we usually adjust the range of f0 for each speaker.
> In the case of ASR, it is difficult. So I'm not sure what will be happened.

I think the unadjusted range would be working as a sort of considering the speaker/environment characteristics, and it would be good.
I want to use a pythonic pitch implementation, so I'll try it.

sw005320 commented on espnet2/tts/feats_extract/energy.py:
```diff
@@ -0,0 +1,134 @@
+# Copyright 2020 Nagoya University (Tomoki Hayashi)
```
This function might be duplicated in ASR and SE, but it has some unique methods (duration and frame length adjustment), and we can leave it as it is.

sw005320 commented on egs2/csmsc/tts1/conf/tuning/decode_tacotron2.yaml:
```diff
@@ -7,9 +7,10 @@
 ##########################################################
```
Should ASR/SE follow a similar style?
I like this style.

sw005320 commented on egs2/jsut/tts1/conf/tuning/train_fastspeech.yaml:
```diff
@@ -0,0 +1,80 @@
+# It requires only a single GPU with 12 GB memory and it
+# takes ~2 days to finish the training on Titan V.
+
+# As default, we assume that the training is based on knowledge
```
```suggestion
# As a default, we assume that the training is based on knowledge
```

kan-bayashi replied:
Sorry for many typo lol

sw005320 commented on espnet2/tts/feats_extract/dio.py:
```diff
@@ -23,7 +23,22 @@
+        applications`: https://doi.org/10.1587/transinf.2015EDP7457
+
+    Note:
+        This module based on numpy implementaion. Therefore, the computational graph
```
```suggestion
        This module is based on NumPy implementation. Therefore, the computational graph
```

kan-bayashi replied:
Sorry for many typo lol

r9y9 commented on espnet2/tts/fastspeech2.py:
```diff
@@ -0,0 +1,680 @@
+        energy_predictor_dropout: float = 0.5,
+        energy_embed_kernel_size: int = 9,
+        energy_embed_dropout: float = 0.5,
+        stop_gradient_from_energy_predictor: bool = False,
```
I am not sure if this works well. Did you find it useful? 

I guess you are concerned about https://github.com/espnet/espnet/issues/2019#issuecomment-657189302. If the problem is the overfitting of variance adapter modules, I would turn loss weights and dropout rates.

kan-bayashi replied:
This option is derived from the advice from one of the authors.
Now I'm checking the effectiveness.

r9y9 replied:
Oops, I forgot about that. Thanks for the explanation.

r9y9 commented on espnet2/tts/fastspeech2.py:
```diff
@@ -0,0 +1,680 @@
+        pitch_predictor_dropout: float = 0.5,
+        pitch_embed_kernel_size: int = 9,
+        pitch_embed_dropout: float = 0.5,
+        stop_gradient_from_pitch_predictor: bool = False,
```
Same as above

kan-bayashi commented:
> It seems that there is still room for improvement regarding prosody naturalness. Did you compare samples between 1) FastSpeech2 w/ ground-truth pitch/energy 2) FastSpeech2 w/ predicted pitch/energy? I am curious how close these samples are, and how close the first samples and ground-truth samples regarding prosody.

OK. I will add `use_teacher_forcing` option for FastSpeech.
Then, we can compare them.

r9y9 commented:
To clarify, what I meant by `there is room for improvement` is that there would be possibilities for future research to further enhance prosody.

The FastSpeech2 samples are definitely better than the ones of FastSpeech and that's very nice :+1:

kan-bayashi commented on espnet2/tts/feats_extract/energy.py:
```diff
@@ -0,0 +1,138 @@
+            for start, end in zip(d_cumsum[:-1], d_cumsum[1:])
+        ]
+        return torch.stack(x_avg)
+
```
I found that we do not need to check 0.0 > for energy.
I simplified this part.

kan-bayashi commented on espnet2/tts/variance_predictor.py:
```diff
@@ -0,0 +1,86 @@
+                        kernel_size,
+                        stride=1,
+                        padding=(kernel_size - 1) // 2,
+                    ),
```
In the case of `Conv1d -> ReLU -> LayerNorm`, bias may have slightly effect?

r9y9 replied:
Ah yes, right. I was worried about too much details.

kan-bayashi commented on espnet2/tts/feats_extract/dio.py:
```diff
@@ -0,0 +1,177 @@
+        d_cumsum = F.pad(d.cumsum(dim=0), (1, 0))
+        x_avg = [
+            x[start:end].masked_select(x[start:end].ne(0.0)).mean(dim=0)
+            if len(x[start:end].masked_select(x[start:end].ne(0.0))) != 0
```
Changed.

kan-bayashi commented:
I added teacher forcing option for FastSpeech and FastSpeech2.
And I added samples with teacher forcing.
drive.google.com/drive/folders/18dZz5jP05Vh5bRRK8nT7HnmLXSeLsitm?usp=sharing

If we use groundtruth of pitch, pronunciation is almost perfect.

kan-bayashi commented on egs2/jsut/tts1/conf/tuning/train_fastspeech2.yaml:
```diff
@@ -0,0 +1,101 @@
+    init_type: xavier_uniform         # initialization type
+    init_enc_alpha: 1.0               # initial value of alpha of encoder scaled position encoding
+    init_dec_alpha: 1.0               # initial value of alpha of decoder scaled position encoding
+    transformer_enc_dropout_rate: 0.1            # dropout rate for transformer encoder layer
+    transformer_enc_positional_dropout_rate: 0.1 # dropout rate for transformer encoder positional encoding
+    transformer_enc_attn_dropout_rate: 0.1       # dropout rate for transformer encoder attention layer
+    transformer_dec_dropout_rate: 0.1            # dropout rate for transformer decoder layer
+    transformer_dec_positional_dropout_rate: 0.1 # dropout rate for transformer decoder positional encoding
+    transformer_dec_attn_dropout_rate: 0.1       # dropout rate for transformer decoder attention layer
```
Thanks. I will try dropout rate = 0.2.

kan-bayashi replied:
BTW, I also tried FastPitch setup (Lamb + Warmup + Weight decay) but it did not work well (loss keep the same value from the middle of the training). Maybe we need to tune LR.

r9y9 commented:
For pitch prediction, we may want to try discrete F0 (i.e., 0 as unvoiced regions) + quantization. Quantized F0 modeling has been successfully applied to SPSS and it's worth trying in my opinion. https://ieeexplore.ieee.org/document/8341752

r9y9 commented:
Looking at the FastSpeech 2 paper again, it seems that the authors use quantized one-hot F0 as input, but use scaler F0s as the output and optimize the pitch predictor by MSE loss. Umm, I don't understand why... I think we can use a cross entropy loss instead in the case of quantized F0 modeling.

kan-bayashi commented:
>For pitch prediction, we may want to try discrete F0 (i.e., 0 in unvoiced regions) + quantization. Quantized F0 modeling has been successfully applied to SPSS and it's worth trying in my opinion. ieeexplore.ieee.org/document/8341752

Hmm. In the case of AR model, quantized f0 will be more useful, as we can see in vocoder?

kan-bayashi commented:
> I don't understand why... I think we can use a cross entropy loss instead in the case of quantized F0 modeling.

About quantized f0 modeling, according to https://github.com/espnet/espnet/issues/2019#issuecomment-657173029, the accuracy seems not so high.

dathudeptrai commented:
@kan-bayashi @r9y9 for FastSpeech2 paper, they use continuous f0 to train with MSE loss and use one-hot F0 as input to feed into the model. In the inference, they use real-value F0 outputs then transform it to one-hot f0 then fit it into the models. Use crossentropy is this case not make sense because the cross entropy (or accuracy) tell not thing about the variant of f0. It just like, groundtruth is category 1 but predict is category 200 or 2 has accuracy 0 %. But all we know that, 2 is better than 200.

kan-bayashi commented:
@dathudeptrai Thank you for your explanation.
I agree that MSE and MAE can consider the distance (how far from the ground truth).

kan-bayashi commented:
I think this PR is ready to merge.
I will merge this evening.

sw005320 commented:
Cool!
